core/src/main/java/org/mule/runtime/core/internal/processor/AsyncDelegateMessageProcessor.java

```diff
@@ -335,22 +341,36 @@ public void handleBackpressure(CoreEvent event) {
       asyncQueue.offer(event);
     }
 
+    public void asyncTaskFinished() {
+      if (scheduledDrain == null) {
+        synchronized (this) {
+          if (scheduledDrain == null) {
+            scheduledDrain = queueDispatcherScheduler.schedule(() -> {
+              while (asyncQueue.size() > 0) {
+                try {
+                  final CoreEvent event = asyncQueue.peek();
+                  if (event != null) {
+                    eventDispatcher.accept(event);
+                    asyncQueue.remove(event);
+                  }
+                } catch (FromFlowRejectedExecutionException free) {
+                  synchronized (this) {
+                    scheduledDrain = null;
+                  }
+                  // nothing to do, a retry will come next
+                  return;
+                }
+              }
+            }, 200, MILLISECONDS);
```

Comment:
why do we still need to keep this polling every 200 ms? If this tasks gets scheduled because of a brief high demand lapse, then this task will continue to run every 200ms without need. Can't we use a traditional single threaded wait/notify()?

Reply:
the thing is that the wait/notify mechanism will still keep a thread busy waiting for elements to arrive at the queue.
With this approach, this task is active only while it makes sense to poll the queue.